Code/GraphMol/ForceFieldHelpers/Wrap/testHelpers.py

```diff
@@ -135,249 +146,324 @@ def test7(self):
+        grad2 = ff.CalcGrad()
+        for v in grad2:
+            self.assertAlmostEqual(v, 0.0, 3)
+        ff.Initialize()
+        grad2 = ff.CalcGrad(savedPos)
+        self.assertEqual(len(grad1), len(grad2))
+        for i in range(len(grad1)):
+            self.assertAlmostEqual(grad1[i], grad2[i], 3)
+
+    def testGitHub2820(self):
+        m = Chem.MolFromSmiles("[Na]C")
+        self.assertIsNotNone(m)
+        mp = ChemicalForceFields.MMFFGetMoleculeProperties(m)
+        self.assertIsNone(mp)
+        rdDistGeom.EmbedMultipleConfs(m, 2)
+        res = ChemicalForceFields.MMFFOptimizeMoleculeConfs(m)
+        self.assertEqual(len(res), 2)
+        self.assertEqual(res[0], res[1])
+        self.assertEqual(res[0], (-1, -1.0))
+
+    def testOptimizeMolecule(self):
+        m = Chem.AddHs(Chem.MolFromSmiles("CCCO"))
+        self.assertIsNotNone(m)
+        self.assertEqual(rdDistGeom.EmbedMolecule(m), 0)
+        mp = ChemicalForceFields.MMFFGetMoleculeProperties(m)
+        ff = ChemicalForceFields.MMFFGetMoleculeForceField(m, mp)
+        before = ff.CalcEnergy()
+        self.assertEqual(ChemicalForceFields.OptimizeMolecule(ff, maxIters=200), 0)
+        after = ff.CalcEnergy()
+        self.assertLess(after, before)
+
+    def testOptimizeMoleculeConfs(self):
+        m = Chem.AddHs(Chem.MolFromSmiles("CCCO"))
+        self.assertIsNotNone(m)
+        cids = rdDistGeom.EmbedMultipleConfs(m, numConfs=10)
+        self.assertEqual(len(cids), 10)
+        mp = ChemicalForceFields.MMFFGetMoleculeProperties(m)
+        ff = ChemicalForceFields.MMFFGetMoleculeForceField(m, mp)
+        before = [
+            ChemicalForceFields.MMFFGetMoleculeForceField(
+                m, mp, confId=cid
+            ).CalcEnergy()
+            for cid in cids
+        ]
+        res, after = tuple(
+            zip(*ChemicalForceFields.OptimizeMoleculeConfs(m, ff, maxIters=200))
+        )
+        self.assertEqual(len(res), 10)
+        self.assertEqual(len(before), len(after))
+        self.assertTrue(all(map(lambda i: i == 0, res)))
+        self.assertTrue(all(after[i] < b for i, b in enumerate(before)))
+
+    def testEmptyFF(self) -> None:
+        m = Chem.AddHs(Chem.MolFromSmiles("CCCO"))
+        rdDistGeom.EmbedMolecule(m)
+        self.assertIsNotNone(m)
+        ff = ChemicalForceFields.GetEmptyForceField(m)
+        self.assertTrue(ff)
+        self.assertFalse(ff.Initialize())
+        self.assertEqual(ff.CalcEnergy(), 0.0)
+        self.assertTrue(all(v == 0.0 for v in ff.CalcGrad()))
+        self.assertEqual(ff.NumPoints(), m.GetNumAtoms())
+        self.assertEqual(len(ff.Positions()) / 3, m.GetNumAtoms())
+        self.assertFalse(ff.Minimize())
+
+        ff.MMFFAddDistanceConstraint(0, 1, False, 100, 100, 100)
+        self.assertFalse(ff.Minimize())
+        pos = ff.Positions()
+        dist = (
+            (pos[0] - pos[3]) ** 2 + (pos[1] - pos[4]) ** 2 + (pos[2] - pos[5]) ** 2
+        ) ** 0.5
```

Comment:
I think it makes more sense to check the distances using the points from the conformer as you do below in the C++ test. this not only tests that the constraints are being obeyed, but also ensures that the points from the conformer aren't just being copied to the force field.

You can use `(p1 - p2).Length()` here too.

Reply:
yes. But there is no other way than to calculate the distance for the force fields position as i did right? the pyforcefield only gives me access to all the positions at once i belive?

Reply:
> yes. But there is no other way than to calculate the distance for the force fields position as i did right? the pyforcefield only gives me access to all the positions at once i belive?

yes, that's correct